src/main/scala/dev/kamu/cli/utility/DependencyGraph.scala

```diff
@@ -0,0 +1,25 @@
+package dev.kamu.cli.utility
+
+class DependencyGraph[A](dependsOn: A => List[A]) {
+  // TODO: Object creation should fail if the graph contains loops and report the loop as an error
+
+  def resolve(goal: List[A]): Seq[A] = {
+    type State = (Seq[A], Set[A]) // (dataset queue, queued datasets)
+    def add(x: A)(s: State): State = (s._1 :+ x, s._2 + x)
+    def isQueued(s: State)(x: A): Boolean = s._2.contains(x)
+    val emptyState = (Seq.empty[A], Set.empty[A])
+
+    @scala.annotation.tailrec
+    def queue(s: State, stack: List[A]): State = {
+      if (stack.nonEmpty) {
+        val deps = dependsOn(stack.head).reverse.filterNot(isQueued(s))
```

Comment:
what's the point of reversing the `dependsOn` here? all dependencies are equivalent, there's no meaningful order between them...

Reply:
There is meaningful order for for the graph, so it produces vertices in post-order order, which is easy to test. Yet reserve was indeed unnecessary here since it gets reversed again later.